src/Datadog.Trace.ClrProfiler.Managed/Integrations/Caching/InterceptedMethodCache.cs

```diff
@@ -0,0 +1,48 @@
+using System;
+using System.Collections.Concurrent;
+using System.Reflection;
+
+namespace Datadog.Trace.ClrProfiler.Integrations
+{
+    /// <summary>
+    /// Wrapper class for retrieving instrumented methods and caching them.
+    /// Necessary because our profiling implementation treats all non <see cref="ValueType"/> types as equal for method signature purposes.
+    /// Uses the actual types of arguments from the calling method to differentiate between signatures and caches them.
+    /// </summary>
+    /// <typeparam name="TDelegate">The type of delegate being intercepted.</typeparam>
+    internal class InterceptedMethodCache<TDelegate>
+        where TDelegate : Delegate
+    {
+        private readonly ConcurrentDictionary<string, TDelegate> _methodCache = new ConcurrentDictionary<string, TDelegate>();
+
+        internal TDelegate GetInterceptedMethod(
```

Comment:
When the function is called the first time, we check the cache to see if we have our wrapper in there.

if it isn't in there, we build the wrapper and add to the cache, then return.

if it is in there, then we return it.

if so, this method might deserve a nice little doc string. The part that confused me when looking at it was the bulk of this function if focused on building and adding something to the cache, when it is called `GetInterceptedMethod`, I really expected it to only return something from the cache if it existed.

Reply:
Perhaps then, `Cache` is not a good name for this class?
This is more a general access or repository class I suppose.
`if not in cache get from expensive operation and store in cache then return`

Reply:
yeah, `Cache` still seems like it makes sense... not sure my comment actually warrants any kind of change, just perspective as an outside reader scanning the code.